File: dotnet/test/MyDotey.SCF.SimpleTest/Source/StringProperty/Cascaded/KeyCachedCascadedConfigurationSource.cs
using System;

using MyDotey.Collections.Generic;
using MyDotey.SCF.Facade;

namespace MyDotey.SCF.Source.StringProperty.Cascaded
{
    /**
     * @author koqizhao
     *
     * Jul 20, 2018
     * 
     * use CascadedKeyDictionary to cache the keys so as to prevent temp string creation and have better young gc
     */
    public class KeyCachedCascadedConfigurationSource<C> : CascadedConfigurationSource<C>
        where C : ConfigurationSourceConfig
    {
        private CascadedKeyDictionary<string, string> _cascadedKeyDictionary;

        public KeyCachedCascadedConfigurationSource(CascadedConfigurationSourceConfig<C> config)
            : base(config)
        {
            _cascadedKeyDictionary = new CascadedKeyDictionary<string, string>();
        }

        protected override string GetKey(params string[] keyParts)
        {
            return _cascadedKeyDictionary.GetOrAdd(base.GetKey, keyParts);
        }
    }
}

File: java/src/test/java/org/mydotey/scf/source/cascaded/KeyCachedCascadedConfigurationSource.java
package org.mydotey.scf.source.cascaded;

import org.mydotey.collection.CascadedKeyMap;

/**
 * @author koqizhao
 *
 * Jul 20, 2018
 * 
 * use CascadedKeyMap to cache the keys so as to prevent temp string creation and have better young gc
 */
public class KeyCachedCascadedConfigurationSource
    extends CascadedConfigurationSource {

    private CascadedKeyMap<String, String> _cascadedKeyMap;

    public KeyCachedCascadedConfigurationSource(CascadedConfigurationSourceConfig config) {
        super(config);

        _cascadedKeyMap = new CascadedKeyMap<>();
    }

    @Override
    protected String getKey(String... keyParts) {
        return _cascadedKeyMap.getOrAdd(super::getKey, keyParts);
    }

}
